### PR TITLE
Fix diagnostic push/pop in C++

### LINKED ISSUE #2962: Review deprecation warnings work as expected in C++
We currently are not getting warnings with Clang and GCC when called proxy operations that are marked as deprecated in Slice.

I did some testing and we get the expected deprecated warnings if we comment out the pus/pop disable warnings in the generated headers.

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR fixes #2962.

The main bug was in SlicedData.h, which was missing a pop.

### CODE REVIEW

File: cpp/test/IceGrid/allocation/AllTests.cpp [hunk 1]
```diff
@@ -126,7 +126,10 @@ class StressClient
             }
 
             assert(session);
-            session->ice_ping();
+#include <Ice/PushDisableWarnings.h>
```
Comment (bernardnormier) [inline cpp/test/IceGrid/allocation/AllTests.cpp:129]: You can remove these 2 include to verify you get a warning.

File: cpp/src/ice2slice/Gen.cpp [hunk 2]
```diff
@@ -6,12 +6,6 @@
 
 #include <cassert>
 
-#if defined(__clang__)
```
Comment (bernardnormier) [inline cpp/src/ice2slice/Gen.cpp:9]: Not used.

File: cpp/src/IceGrid/SessionManager.h [hunk 3]
```diff
@@ -16,20 +16,13 @@ namespace IceGrid
 {
     template<class TPrx> class SessionKeepAliveThread
     {
-#if defined(__clang__)
```
Comment (bernardnormier) [inline cpp/src/IceGrid/SessionManager.h:19]: Not used.

File: cpp/src/IceGrid/ServerI.h [hunk 4]
```diff
@@ -29,10 +29,6 @@ namespace IceGrid
     class ServerI final : public Server, public std::enable_shared_from_this<ServerI>
     {
     public:
-#if defined(__clang__)
```
Comment (bernardnormier) [inline cpp/src/IceGrid/ServerI.h:32]: Not used.

File: cpp/src/DataStorm/TraceUtil.cpp [hunk 5]
```diff
@@ -7,6 +7,14 @@
 using namespace std;
 using namespace DataStormI;
 
+#if defined(__clang__)
```
Comment (bernardnormier) [inline cpp/src/DataStorm/TraceUtil.cpp:10]: Somehow this was relying on the SlicedData.h push.

File: cpp/src/DataStorm/SessionI.h [hunk 6]
```diff
@@ -9,13 +9,12 @@
...
 #include "NodeI.h"
 
-#if defined(_MSC_VER)
-#    pragma warning(push)
-#    pragma warning(disable : 4456) // ... : declaration of 'identifier' hides previous local declaration
-#    pragma warning(disable : 4458) // ... : declaration of 'identifier' hides class member
-#elif defined(__clang__)
+#if defined(__clang__)
 #    pragma clang diagnostic push
-#    pragma clang diagnostic ignored "-Wshadow"
```
Comment (bernardnormier) [inline cpp/src/DataStorm/SessionI.h:18]: We should never disable plain -Wshadow, especially not in a header file. I fixed the shadowing in this PR.

File: cpp/src/DataStorm/NodeSessionI.cpp [hunk 7]
```diff
@@ -33,8 +33,8 @@ namespace
             {
                 try
                 {
-                    optional<SessionPrx> session;
-                    updateNodeAndSessionProxy(*publisher, session, current);
+                    optional<SessionPrx> sessionPrx;
```
Comment (bernardnormier) [inline cpp/src/DataStorm/NodeSessionI.cpp:36]: Fixed shadowing.

**Review by pepone — APPROVED** [review]

**Pull request merged**